tools/codesign_mac_executable.py

```diff
@@ -0,0 +1,232 @@
+
+    for file in files:
+        cmd = ["sudo", "codesign", "-f", "-s", "-v", "--options", "runtime", "--timestamp", "--entitlements", args.entitlements_file, "--sign", f"Developer ID Application: {args.developer_id}", file]
+        print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+        cmd_result = subprocess.run(cmd)
+        assert cmd_result.returncode == 0
+
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    # create a zip file for notarization
+    temp_zip = os.path.join(args.output_dir, f"{os.path.splitext(os.path.basename(args.macos_executable))[0]}-temp.zip")
+    cmd = ["ditto", "-c", "-k", "--rsrc", "--keepParent", args.macos_executable, temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    cmd_result = subprocess.run(cmd)
+    assert cmd_result.returncode == 0
+
+    # send the zip file for notarization 
+    cmd = ["xcrun", "altool", "--notarize-app", "--primary-bundle-id", "org.embodiedaifoundation.spear", "--username", args.apple_username, "--password", args.apple_password, "--file", temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+    output = subprocess.check_output(["grep", "RequestUUID = "], stdin=ps.stdout, text=True)
+    ps.wait()
+    ps.stdout.close()
+    request_uuid = " ".join(output[14:].split())
+    print(f"[SPEAR | codesign_mac_executable.py] Request UUID: {request_uuid}")
+
+    # check the status of notarization request
+    cmd = [
+        "xcrun", "altool", "--notarization-info", request_uuid, "--username", args.apple_username, "--password", args.apple_password]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    output = "             in progress"
+    start_time = time.time()
+    elapsed_time = time.time() - start_time
+    while " ".join(output[13:].split()) == "in progress" and elapsed_time<args.wait_time_seconds:
+        ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        output = subprocess.check_output(["grep", "     Status:"], stdin=ps.stdout, text=True)
+        ps.wait()
+        ps.stdout.close()
+        time.sleep(20)
+        elapsed_time = time.time() - start_time
+        print(f"[SPEAR | codesign_mac_executable.py] Waiting to get more information on the notarization request...")
+    
+    assert elapsed_time <= args.wait_time_seconds, f"[SPEAR | codesign_mac_executable.py] ERROR: Exceeded provided wait time ({args.wait_time_seconds}s) for notarization. Please complete this procedure after notarization is complete."
+    
+    # staple the executable
+    cmd = ["xcrun", "stapler", "staple", args.macos_executable]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    cmd_result = subprocess.run(cmd)
+    assert cmd_result.returncode == 0
+
+    print(f"[SPEAR | codesign_mac_executable.py] Removing the temporary zip file {temp_zip}") 
```

Comment:
don't remove, just write to a user-specifiable temp folder

Reply:
Done